File: rust/examples/19585_v1.rs
// 전설

use std::collections::{HashMap, HashSet};
use std::error::Error;
use std::fmt::Write;
use std::io::{stdin, Read};

struct Trie {
    tree: HashMap<u8, Trie>,
    is_word: bool,
}

impl Trie {
    fn insert(&mut self, word: &[u8]) {
        if word.is_empty() {
            return;
        }
        let entry = self.tree.entry(word[0]).or_insert_with(|| Trie {
            tree: HashMap::new(),
            is_word: false,
        });
        if word.len() == 1 {
            entry.is_word = true;
        }
        entry.insert(&word[1..]);
    }

    fn find(&self, word: &[u8], idx: usize, result: &mut Vec<usize>) {
        if self.is_word {
            // 단어라서 넣음
            result.push(idx - 1);
        }

        if word.is_empty() {
            return;
        }

        if let Some(next_trie) = self.tree.get(&word[0]) {
            next_trie.find(&word[1..], idx + 1, result);
        }
    }
}

fn main() -> Result<(), Box<dyn Error>> {
    let mut output = String::new();
    let mut input = String::new();
    stdin().read_to_string(&mut input).unwrap();
    let mut input = input.split_ascii_whitespace();

    let num_c = input.next().unwrap().parse::<usize>()?;
    let num_n = input.next().unwrap().parse::<usize>()?;

    let mut color_trie = Trie {
        is_word: false,
        tree: HashMap::new(),
    };

    for color in input.by_ref().take(num_c).map(|x| x.as_bytes()) {
        color_trie.insert(color);
    }

    let nicknames: Vec<&[u8]> = input.by_ref().take(num_n).map(|x| x.as_bytes()).collect();

    let nickname_set: HashSet<&[u8]> = nicknames.into_iter().collect();

    let num_q = input.next().unwrap().parse::<usize>()?;

    for _ in 0..num_q {
        let team = input.next().unwrap().as_bytes();

        let mut color_result = vec![];

        color_trie.find(team, 0, &mut color_result);
        let mut result = false;

        for cdx in color_result {
            if nickname_set.contains(&team[(cdx + 1)..]) {
                result = true;
                break;
            }
        }
        if !result {
            writeln!(output, "NO")?;
        } else {
            writeln!(output, "YES")?;
        }
    }
    print!("{}", output);
    Ok(())
}

File: python3/19585_v1.py
# 전설

from sys import stdin

read = lambda: stdin.readline().rstrip()


class Trie:
    def __init__(self, is_word):
        self.tree = {}
        self.is_word = is_word

    def insert(self, word: str):
        curr_node = self
        for char in word:
            if char not in curr_node.tree:
                curr_node.tree[char] = Trie(False)
            curr_node = curr_node.tree[char]
        curr_node.is_word = True

    def find(self, word: str, result: list[int]):
        # 닉네임에서 색 이름이 아니게 되는 인덱스를 result 에 넣기
        curr_node = self
        for idx, char in enumerate(word):
            if curr_node.is_word:
                result.append(idx - 1)
            if char in curr_node.tree:
                curr_node = curr_node.tree[char]
            else:
                break


num_c, num_n = map(int, read().split())

color_trie = Trie(False)

for idx in range(num_c):
    color = read()
    color_trie.insert(color)

nickname_set = set()

for idx in range(num_n):
    nickname = read()
    nickname_set.add(nickname)

num_q = int(read())

for _ in range(num_q):
    team = read()

    color_result = []
    color_trie.find(team, color_result)

    result = False

    for cdx in color_result:
        remaining = team[cdx + 1 :]
        if remaining in nickname_set:
            result = True
            break

    if not result:
        print("No")
    else:
        print("Yes")
